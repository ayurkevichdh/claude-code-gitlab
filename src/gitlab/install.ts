import { readFile } from "fs/promises";
import { resolve } from "path";
import { Gitlab } from "@gitbeaker/rest";

export async function upsertVariables(
  api: Pick<Gitlab, "ProjectVariables">,
  projectId: string | number,
  gitlabToken: string,
  anthropicKey: string,
) {
  async function upsert(key: string, value: string) {
    try {
      await (api as any).ProjectVariables.create(projectId, key, value);
      console.log(`Created variable ${key}`);
    } catch (_) {
      await (api as any).ProjectVariables.edit(projectId, key, value);
      console.log(`Updated variable ${key}`);
    }
  }

  await upsert("GITLAB_TOKEN", gitlabToken);
  await upsert("ANTHROPIC_API_KEY", anthropicKey);
}

export async function ensureCiFile(api: Gitlab, projectId: string | number) {
  const project = await api.Projects.show(projectId as any);
  const branch = project.default_branch as string;
  const filePath = ".gitlab-ci.yml";
  try {
    await api.RepositoryFiles.show(projectId as any, filePath, branch);
    console.log(`${filePath} already exists`);
  } catch (_) {
    const content = await readFile(resolve("examples/gitlab-ci.yml"), "utf8");
    await api.RepositoryFiles.create(
      projectId as any,
      filePath,
      branch,
      content,
      "Add Claude CI",
    );
    console.log(`Committed ${filePath} to ${branch}`);
  }
}
