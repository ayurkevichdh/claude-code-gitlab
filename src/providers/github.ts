import { $ } from "bun";
import type { Octokits } from "../github/api/client";
import type { ParsedGitHubContext } from "../github/context";
import type { IProvider } from "./IProvider";

export class GitHubProvider implements IProvider {
  constructor(
    private octokits: Octokits,
    private context: ParsedGitHubContext,
  ) {}

  async getDiff(
    base: string,
    head: string,
    filePath?: string,
  ): Promise<string> {
    const args = [base, head];
    if (filePath) {
      args.push("--", filePath);
    }
    const { stdout } = await $`git diff ${args}`.quiet();
    return stdout.toString();
  }

  async createProgressComment(body: string): Promise<number> {
    const { owner, repo } = this.context.repository;
    const res = await this.octokits.rest.issues.createComment({
      owner,
      repo,
      issue_number: this.context.entityNumber,
      body,
    });
    return res.data.id;
  }

  async updateComment(commentId: number, body: string): Promise<void> {
    const { owner, repo } = this.context.repository;
    await this.octokits.rest.issues.updateComment({
      owner,
      repo,
      comment_id: commentId,
      body,
    });
  }

  async addInlineComment(
    filePath: string,
    line: number,
    body: string,
  ): Promise<number> {
    const { owner, repo } = this.context.repository;
    const pr = await this.octokits.rest.pulls.get({
      owner,
      repo,
      pull_number: this.context.entityNumber,
    });
    const res = await this.octokits.rest.pulls.createReviewComment({
      owner,
      repo,
      pull_number: this.context.entityNumber,
      commit_id: pr.data.head.sha,
      body,
      path: filePath,
      line,
    });
    return res.data.id;
  }

  async pushFixupCommit(message: string): Promise<string> {
    await $`git add -A`.quiet();
    await $`git commit -m ${message}`.quiet();
    const { stdout } = await $`git rev-parse HEAD`.quiet();
    const sha = stdout.toString().trim();
    await $`git push`.quiet();
    return sha;
  }
}
