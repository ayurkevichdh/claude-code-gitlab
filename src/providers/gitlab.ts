import { $ } from "bun";
import fetch from "node-fetch";
import type { IProvider } from "./IProvider";

export interface GitLabContext {
  projectId: string;
  mergeRequestIid: number;
}

export class GitLabProvider implements IProvider {
  constructor(
    private token: string,
    private context: GitLabContext,
    private baseUrl = "https://gitlab.com/api/v4",
  ) {}

  async getDiff(
    base: string,
    head: string,
    filePath?: string,
  ): Promise<string> {
    const args = [base, head];
    if (filePath) args.push("--", filePath);
    const { stdout } = await $`git diff ${args}`.quiet();
    return stdout.toString();
  }

  async createProgressComment(body: string): Promise<number> {
    const url = `${this.baseUrl}/projects/${encodeURIComponent(this.context.projectId)}/merge_requests/${this.context.mergeRequestIid}/notes`;
    const res = await fetch(url, {
      method: "POST",
      headers: {
        "PRIVATE-TOKEN": this.token,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ body }),
    });
    const data = (await res.json()) as { id: number };
    return data.id;
  }

  async updateComment(commentId: number, body: string): Promise<void> {
    const url = `${this.baseUrl}/projects/${encodeURIComponent(this.context.projectId)}/merge_requests/${this.context.mergeRequestIid}/notes/${commentId}`;
    await fetch(url, {
      method: "PUT",
      headers: {
        "PRIVATE-TOKEN": this.token,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ body }),
    });
  }

  async addInlineComment(
    filePath: string,
    line: number,
    body: string,
  ): Promise<number> {
    const url = `${this.baseUrl}/projects/${encodeURIComponent(this.context.projectId)}/merge_requests/${this.context.mergeRequestIid}/discussions`;
    const res = await fetch(url, {
      method: "POST",
      headers: {
        "PRIVATE-TOKEN": this.token,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        body,
        position: {
          position_type: "text",
          new_path: filePath,
          new_line: line,
        },
      }),
    });
    const data = (await res.json()) as { notes?: { id: number }[] };
    return data.notes?.[0]?.id ?? 0;
  }

  async pushFixupCommit(message: string): Promise<string> {
    await $`git add -A`.quiet();
    await $`git commit -m ${message}`.quiet();
    const { stdout } = await $`git rev-parse HEAD`.quiet();
    const sha = stdout.toString().trim();
    await $`git push`.quiet();
    return sha;
  }
}
