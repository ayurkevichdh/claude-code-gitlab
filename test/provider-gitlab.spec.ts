import { describe, it, expect, beforeEach, afterEach, spyOn } from "bun:test";
import * as nodeFetch from "node-fetch";
import { GitLabProvider } from "../src/providers/gitlab";

const context = { projectId: "123", mergeRequestIid: 10 };

describe("GitLabProvider", () => {
  let fetchSpy: any;

  beforeEach(() => {
    fetchSpy = spyOn(nodeFetch, "default").mockResolvedValue({
      json: async () => ({ id: 99, notes: [{ id: 88 }] }),
    } as any);
  });

  afterEach(() => {
    fetchSpy.mockRestore();
  });

  it("creates progress comment", async () => {
    const provider = new GitLabProvider("TOKEN", context);
    const id = await provider.createProgressComment("Work started");

    expect(fetchSpy).toHaveBeenCalledWith(
      "https://gitlab.com/api/v4/projects/123/merge_requests/10/notes",
      {
        method: "POST",
        headers: {
          "PRIVATE-TOKEN": "TOKEN",
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ body: "Work started" }),
      },
    );
    const body = JSON.parse((fetchSpy.mock.calls[0][1] as any).body);
    expect(body).toMatchSnapshot();
    expect(id).toBe(99);
  });

  it("retrieves diff between refs", async () => {
    const provider = new GitLabProvider("TOKEN", context);
    const diff = await provider.getDiff("HEAD", "HEAD");
    expect(diff).toMatchSnapshot();
  });
});
