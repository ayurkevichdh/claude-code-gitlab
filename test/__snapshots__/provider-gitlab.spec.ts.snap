// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`GitLabProvider retrieves diff between refs 1`] = `""`;

exports[`GitLabProvider creates progress comment 1`] = `
{
  "body": "Work started",
}
`;
